File: node-mcp/src/searxng.ts
import { SEARXNG_URL } from "./config.js";

export interface SearxngResult {
  title: string;
  url: string;
  content: string;
  score?: number;
  category?: string;
  publishedDate?: string;
  engine?: string;
  engines?: string[];
}

export interface SearxngResponse {
  query: string;
  results: SearxngResult[];
  answers?: string[];
  infoboxes?: Array<{ infobox: string; content: string }>;
  suggestions?: string[];
  number_of_results?: number;
}

export async function searchSearxng(
  query: string,
  options: {
    categories?: string;
    language?: string;
    numResults?: number;
    timeRange?: string;
  } = {}
): Promise<SearxngResponse> {
  const params = new URLSearchParams({
    q: query,
    format: "json",
    categories: options.categories ?? "general",
    ...(options.language && { language: options.language }),
    ...(options.timeRange && { time_range: options.timeRange }),
  });

  const response = await fetch(`${SEARXNG_URL}/search?${params}`, {
    headers: {
      "Accept": "application/json",
      "Accept-Language": "en-US,en;q=0.9",
      "User-Agent": "Mozilla/5.0 (compatible; SearXNG-MCP/1.0)",
    },
  });

  if (!response.ok) {
    throw new Error(`SearXNG returned ${response.status}: ${response.statusText}`);
  }

  const data = (await response.json()) as SearxngResponse;

  if (options.numResults) {
    data.results = data.results.slice(0, options.numResults);
  }

  return data;
}

export function formatResults(data: SearxngResponse): string {
  const lines: string[] = [];

  if (data.answers && data.answers.length > 0) {
    lines.push(`**Direct answer:** ${data.answers[0]}\n`);
  }

  if (data.infoboxes && data.infoboxes.length > 0) {
    lines.push(`**${data.infoboxes[0].infobox}:** ${data.infoboxes[0].content}\n`);
  }

  if (data.results.length === 0) {
    return "No results found.";
  }

  data.results.forEach((result, i) => {
    lines.push(`${i + 1}. **${result.title}**`);
    lines.push(`   URL: ${result.url}`);
    if (result.content) {
      lines.push(`   ${result.content}`);
    }
    if (result.publishedDate) {
      lines.push(`   Published: ${result.publishedDate}`);
    }
    lines.push("");
  });

  if (data.suggestions && data.suggestions.length > 0) {
    lines.push(`\n**Related searches:** ${data.suggestions.slice(0, 5).join(", ")}`);
  }

  return lines.join("\n");
}

File: py-mcp/src/searxng.py
"""SearXNG API integration."""

from dataclasses import dataclass
from typing import Any
import urllib.request
import urllib.parse
import json

from .config import SEARXNG_URL, DEFAULT_LANGUAGE


@dataclass
class SearxngResult:
    title: str
    url: str
    content: str
    published_date: str | None = None
    engine: str | None = None
    engines: list[str] | None = None


@dataclass
class SearxngResponse:
    query: str
    results: list[SearxngResult]
    answers: list[Any] | None = None
    infoboxes: list[dict[str, str]] | None = None
    suggestions: list[str] | None = None


def search_searxng(
    query: str,
    *,
    categories: str = "general",
    language: str | None = None,
    num_results: int = 10,
    time_range: str | None = None,
) -> SearxngResponse:
    """Search SearXNG and return parsed results."""
    params = {
        "q": query,
        "format": "json",
        "categories": categories,
    }
    if language:
        params["language"] = language
    if time_range:
        params["time_range"] = time_range

    url = f"{SEARXNG_URL}/search?{urllib.parse.urlencode(params)}"
    headers = {
        "Accept": "application/json",
        "Accept-Language": "en-US,en;q=0.9",
        "User-Agent": "Mozilla/5.0 (compatible; SearXNG-MCP/1.0)",
    }

    req = urllib.request.Request(url, headers=headers, method="GET")
    with urllib.request.urlopen(req) as resp:
        if resp.status != 200:
            raise Exception(f"SearXNG returned {resp.status}")
        data = json.loads(resp.read().decode("utf-8"))

    results = [
        SearxngResult(
            title=r.get("title", ""),
            url=r.get("url", ""),
            content=r.get("content", ""),
            published_date=r.get("publishedDate"),
            engine=r.get("engine"),
            engines=r.get("engines"),
        )
        for r in data.get("results", [])[:num_results]
    ]

    return SearxngResponse(
        query=data.get("query", ""),
        results=results,
        answers=data.get("answers"),
        infoboxes=data.get("infoboxes"),
        suggestions=data.get("suggestions"),
    )


def format_results(data: SearxngResponse) -> str:
    """Format SearXNG response as markdown text."""
    lines: list[str] = []

    if data.answers:
        answer = data.answers[0]
        if isinstance(answer, dict):
            answer = answer.get("answer", "")
        lines.append(f"**Direct answer:** {answer}\n")

    if data.infoboxes:
        infobox = data.infoboxes[0]
        lines.append(f"**{infobox.get('infobox', 'Info')}:** {infobox.get('content', '')}\n")

    if not data.results:
        return "No results found."

    for i, r in enumerate(data.results, 1):
        lines.append(f"{i}. **{r.title}**")
        lines.append(f"   URL: {r.url}")
        if r.content:
            lines.append(f"   {r.content}")
        if r.published_date:
            lines.append(f"   Published: {r.published_date}")
        lines.append("")

    if data.suggestions:
        lines.append(f"\n**Related searches:** {', '.join(data.suggestions[:5])}")

    return "\n".join(lines)